File: core/src/main/java/y2k/joyreactor/services/requests/PostRequest.java
package y2k.joyreactor.services.requests;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import y2k.joyreactor.*;
import y2k.joyreactor.http.HttpClient;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by y2k on 11/21/15.
 */
public class PostRequest {

    private static final Pattern POST_ID = Pattern.compile("/post/(\\d+)");

    private Post post;
    private PostCommentsRequest commentsRequest = new PostCommentsRequest();
    private List<SimilarPost> similarPosts = new ArrayList<>();
    private List<Attachment> attachments = new ArrayList<>();

    public List<SimilarPost> getSimilarPosts() {
        return similarPosts;
    }

    public Post getPost() {
        return post;
    }

    public List<Comment> getComments() {
        return commentsRequest.getComments();
    }

    public void request(String postId) throws IOException {
        Document page = HttpClient.getInstance().getDocument(getPostUrl(postId));

        Element postNode = page.select("div.postContainer").first();
        post = PostsForTagRequest.newPost(postNode); // TODO:

        commentsRequest.request(page);

        for (Element e : page.select(".similar_post img")) {
            SimilarPost similarPost = new SimilarPost(getPostId(e.parent().attr("href")));
            similarPost.setImage(new Image(e.absUrl("src"), 0, 0));
            similarPosts.add(similarPost);
        }

        Elements imgElement = postNode.select("div.image > img");
        if (imgElement.size() > 1)
            for (Element e : imgElement.subList(1, imgElement.size() - 1)) {
                Attachment a = new Attachment();
                a.image = new Image(e.absUrl("src"),
                        Integer.parseInt(e.attr("width")),
                        Integer.parseInt(e.attr("height")));
                attachments.add(a);
            }
    }

    private String getPostId(String href) {
        Matcher m = POST_ID.matcher(href);
        if (!m.find()) throw new IllegalStateException();
        return m.group(1);
    }

    private String getPostUrl(String postId) {
        return "http://anime.reactor.cc/post/" + postId; // TODO:
    }

    public List<Attachment> getAttachments() {
        return attachments;
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/requests/PostRequest.kt
package y2k.joyreactor.services.requests

import org.jsoup.nodes.Element
import y2k.joyreactor.common.async.CompletableFuture
import y2k.joyreactor.common.async.runAsync
import y2k.joyreactor.common.http.HttpClient
import y2k.joyreactor.model.*
import java.util.*
import java.util.regex.Pattern

/**
 * Created by y2k on 11/21/15.
 */
class PostRequest(
    private val httpClient: HttpClient,
    private val parser: (Element) -> Pair<Post, List<Attachment>>) :
    Function1<Long, CompletableFuture<PostRequest.Response>> {

    override operator fun invoke(postId: Long) = runAsync { doRequest(postId.toString()) }

    @Deprecated("")
    fun request(postId: String): Response = doRequest(postId)

    private fun doRequest(postId: String): Response {
        val commentsRequest = PostCommentsRequest()
        val similarPosts = ArrayList<SimilarPost>()

        val page = httpClient.getDocument(getPostUrl(postId))

        val postNode = page.select("div.postContainer").first()
        val postWithAttachments = parser(postNode)

        commentsRequest.request(page, postId.toLong())

        for (e in page.select(".similar_post img")) {
            val similarPost = SimilarPost(
                postId.toLong(),
                Image(e.absUrl("src"), 0, 0),
                getPostId(e.parent().attr("href")).toLong())
            similarPosts.add(similarPost)
        }

        return Response(
            postWithAttachments.first, commentsRequest.comments,
            similarPosts, postWithAttachments.second)
    }

    private fun getPostId(href: String): String {
        val m = POST_ID.matcher(href)
        if (!m.find()) throw IllegalStateException()
        return m.group(1)
    }

    private fun getPostUrl(postId: String): String {
        return "http://anime.reactor.cc/post/" + postId // TODO:
    }

    data class Response(
        val post: Post,
        val comments: List<Comment>,
        val similarPosts: List<SimilarPost>,
        val attachments: List<Attachment>)

    companion object {

        private val POST_ID = Pattern.compile("/post/(\\d+)")
    }
}